File: packages/sdkwork-utils-typescript/src/decimal_math.ts
/**
 * Exact decimal arithmetic on integer units.
 *
 * All money / points / token-bank / exchange-rate math that must never lose a
 * fraction is funneled through these helpers. A decimal string is parsed into
 * an integer count of `scale`-th units (e.g. scale 6 => `1.5` -> `1500000`,
 * scale 12 => `0.000704` -> `704000000`), the operation is performed in exact
 * BigInt integer arithmetic, a rounding mode is applied only when a scaling
 * boundary is crossed, and the result is formatted back to a plain decimal
 * string. No `number` (IEEE-754 double) is ever used, so results are exact for
 * arbitrary input magnitude and fractional digits.
 *
 * Rounding modes mirror standard decimal conventions and are only applied at
 * the requested `scale` boundary:
 *  - `"floor"`   truncate toward -infinity
 *  - `"ceil"`    round toward +infinity (billing uses this so a fractional
 *                charge never shortchanges the merchant)
 *  - `"half_up"` round half away from zero
 *  - `"half_even"` banker's rounding (round half to even)
 */

export type DecimalRounding = "floor" | "ceil" | "half_up" | "half_even";

/** The default fractional scale shared by the arithmetic helpers. */
export const DEFAULT_DECIMAL_SCALE = 12;

function normalizeScale(scale: number): number {
  if (!Number.isSafeInteger(scale) || scale < 0 || scale > 38) {
    return DEFAULT_DECIMAL_SCALE;
  }
  return scale;
}

function assertRounding(rounding: DecimalRounding): DecimalRounding {
  if (rounding === "floor" || rounding === "ceil" || rounding === "half_up" || rounding === "half_even") {
    return rounding;
  }
  return "floor";
}

const DECIMAL_PATTERN = /^-?\d+(?:\.\d+)?$/u;

/**
 * Parse a signed decimal string into the integer count of `scale`-th units.
 * A value carrying more fractional digits than `scale` is rounded to `scale`
 * with `rounding`. Returns `null` for empty / malformed input.
 */
export function decimalToScaledUnits(
  value: string,
  scale = DEFAULT_DECIMAL_SCALE,
  rounding: DecimalRounding = "half_up",
): bigint | null {
  const normalizedScale = normalizeScale(scale);
  const mode = assertRounding(rounding);
  const trimmed = value.trim();
  if (!DECIMAL_PATTERN.test(trimmed)) {
    return null;
  }
  const negative = trimmed.startsWith("-");
  const unsigned = negative ? trimmed.slice(1) : trimmed;
  const [wholeRaw = "0", fractionRaw = ""] = unsigned.split(".");
  let whole: bigint;
  try {
    whole = BigInt(wholeRaw);
  } catch {
    return null;
  }
  const n = fractionRaw.length;
  let raw: bigint;
  try {
    raw = whole * 10n ** BigInt(n) + (n > 0 ? BigInt(fractionRaw) : 0n);
  } catch {
    return null;
  }
  let units: bigint;
  if (n <= normalizedScale) {
    units = raw * 10n ** BigInt(normalizedScale - n);
  } else {
    const denominator = 10n ** BigInt(n - normalizedScale);
    units = roundFraction(raw, denominator, mode);
  }
  return negative ? -units : units;
}

/**
 * Parse the exact value of a signed decimal string into integer units without
 * any loss: returns the integer digits and the number of fractional digits.
 * Used internally by arithmetic so the product / quotient stays exact until a
 * final rounding at the requested scale.
 */
function parseExact(value: string): { digits: bigint; scale: number } | null {
  const trimmed = value.trim();
  if (!DECIMAL_PATTERN.test(trimmed)) {
    return null;
  }
  const negative = trimmed.startsWith("-");
  const unsigned = negative ? trimmed.slice(1) : trimmed;
  const [wholeRaw = "0", fractionRaw = ""] = unsigned.split(".");
  let whole: bigint;
  try {
    whole = BigInt(wholeRaw);
  } catch {
    return null;
  }
  const trimmedFraction = fractionRaw.replace(/0+$/u, "");
  let digits = whole * 10n ** BigInt(trimmedFraction.length);
  if (trimmedFraction.length > 0) {
    digits += BigInt(trimmedFraction);
  }
  if (negative) {
    digits = -digits;
  }
  return { digits, scale: trimmedFraction.length };
}

/** Round `num / denom` to an integer using `rounding` (denom must be positive). */
function roundFraction(num: bigint, denom: bigint, rounding: DecimalRounding): bigint {
  if (denom <= 0n) {
    return num;
  }
  const quotient = num / denom;
  const remainder = num % denom;
  if (remainder === 0n) {
    return quotient;
  }
  const absRemainder = remainder < 0n ? -remainder : remainder;
  const doubled = absRemainder * 2n;
  switch (rounding) {
    case "ceil":
      return num > 0n ? quotient + 1n : quotient;
    case "floor":
      return num < 0n ? quotient - 1n : quotient;
    case "half_even":
      if (doubled < denom) return quotient;
      if (doubled > denom) return num > 0n ? quotient + 1n : quotient - 1n;
      return quotient % 2n === 0n ? quotient : quotient + (num > 0n ? 1n : -1n);
    case "half_up":
    default:
      return num > 0n
        ? doubled >= denom ? quotient + 1n : quotient
        : doubled >= denom ? quotient - 1n : quotient;
  }
}

/** Format an integer count of `scale`-th units back into a decimal string, trimming trailing zeros. */
export function scaledUnitsToDecimal(units: bigint, scale = DEFAULT_DECIMAL_SCALE): string {
  const normalizedScale = normalizeScale(scale);
  const negative = units < 0n;
  const absolute = negative ? -units : units;
  const scaleUnit = 10n ** BigInt(normalizedScale);
  const whole = absolute / scaleUnit;
  if (normalizedScale === 0) {
    return negative ? `-${whole}` : whole.toString();
  }
  const fraction = String(absolute % scaleUnit).padStart(normalizedScale, "0").replace(/0+$/u, "");
  const sign = negative ? "-" : "";
  return fraction.length > 0 ? `${sign}${whole}.${fraction}` : `${sign}${whole}`;
}

/**
 * Exact decimal addition of two strings scaled to `scale`, returning a decimal
 * string. Inputs with more fractional digits than `scale` are rounded first.
 */
export function decimalAdd(
  left: string,
  right: string,
  scale = DEFAULT_DECIMAL_SCALE,
  rounding: DecimalRounding = "half_up",
): string {
  const normalizedScale = normalizeScale(scale);
  const mode = assertRounding(rounding);
  const leftUnits = decimalToScaledUnits(left, normalizedScale, mode);
  const rightUnits = decimalToScaledUnits(right, normalizedScale, mode);
  if (leftUnits === null || rightUnits === null) {
    return "0";
  }
  return scaledUnitsToDecimal(leftUnits + rightUnits, normalizedScale);
}

/** Exact decimal subtraction of two strings scaled to `scale`. */
export function decimalSubtract(
  left: string,
  right: string,
  scale = DEFAULT_DECIMAL_SCALE,
  rounding: DecimalRounding = "half_up",
): string {
  const normalizedScale = normalizeScale(scale);
  const mode = assertRounding(rounding);
  const leftUnits = decimalToScaledUnits(left, normalizedScale, mode);
  const rightUnits = decimalToScaledUnits(right, normalizedScale, mode);
  if (leftUnits === null || rightUnits === null) {
    return "0";
  }
  return scaledUnitsToDecimal(leftUnits - rightUnits, normalizedScale);
}

/**
 * Exact decimal multiplication of two strings, returning a decimal string
 * scaled to `scale`. The exact product (at arbitrary combined precision) is
 * computed in BigInt, then rounded to `scale` with `rounding`. Billing uses
 * `("ceil", scale 6)` to mirror `micro = ceil(amount × rate × 1e6)`.
 */
export function decimalMultiply(
  left: string,
  right: string,
  scale = DEFAULT_DECIMAL_SCALE,
  rounding: DecimalRounding = "half_up",
): string {
  const normalizedScale = normalizeScale(scale);
  const mode = assertRounding(rounding);
  const leftExact = parseExact(left);
  const rightExact = parseExact(right);
  if (!leftExact || !rightExact) {
    return "0";
  }
  const product = leftExact.digits * rightExact.digits;
  const productScale = leftExact.scale + rightExact.scale;
  const offset = normalizedScale - productScale;
  let resultUnits: bigint;
  if (offset >= 0) {
    resultUnits = product * 10n ** BigInt(offset);
  } else {
    const denominator = 10n ** BigInt(-offset);
    resultUnits = roundFraction(product, denominator, mode);
  }
  return scaledUnitsToDecimal(resultUnits, normalizedScale);
}

/**
 * Exact decimal division `numerator / denominator`, returning a decimal string
 * scaled to `scale`. A divide-by-zero or malformed input yields `"0"`. The
 * quotient is extended to `scale` fractional digits; when that requires
 * dropping a remainder, `rounding` decides.
 */
export function decimalDivide(
  numerator: string,
  denominator: string,
  scale = DEFAULT_DECIMAL_SCALE,
  rounding: DecimalRounding = "half_up",
): string {
  const normalizedScale = normalizeScale(scale);
  const mode = assertRounding(rounding);
  const numExact = parseExact(numerator);
  const denExact = parseExact(denominator);
  if (!numExact || !denExact || denExact.digits === 0n) {
    return "0";
  }
  // value = num / den
  // resultUnits = numerator / denominator × 10^scale
  //            = (num.digits × 10^scale × 10^den.scale) / (den.digits × 10^num.scale)
  const numPower = normalizedScale + denExact.scale - numExact.scale;
  let resultUnits: bigint;
  if (numPower >= 0) {
    resultUnits = roundFraction(numExact.digits * 10n ** BigInt(numPower), denExact.digits, mode);
  } else {
    const denomScaled = denExact.digits * 10n ** BigInt(-numPower);
    resultUnits = roundFraction(numExact.digits, denomScaled, mode);
  }
  return scaledUnitsToDecimal(resultUnits, normalizedScale);
}

/** Exact summation of decimal strings scaled to `scale` (a sum never needs rounding). */
export function sumDecimalStrings(values: string[], scale = DEFAULT_DECIMAL_SCALE): string {
  const normalizedScale = normalizeScale(scale);
  let total = 0n;
  for (const value of values) {
    const units = decimalToScaledUnits(value, normalizedScale, "floor");
    if (units !== null) {
      total += units;
    }
  }
  return scaledUnitsToDecimal(total, normalizedScale);
}

File: packages/sdkwork-utils-rust/src/decimal_math.rs
//! Exact decimal arithmetic on integer units.
//!
//! Mirror of the TypeScript `@sdkwork/utils` `decimal_math` module so the same
//! add / subtract / multiply / divide never loses a fraction on either side of
//! the wire. A decimal string is parsed into an integer count of `scale`-th
//! units (scale 6 => `1.5` -> `1500000`), the operation is performed in exact
//! `i128` integer arithmetic (checked so overflow is a hard error rather than
//! silent corruption), a rounding mode is applied only when a scaling boundary
//! is crossed, and the result is formatted back to a plain decimal string.
//!
//! Rounding modes mirror the TypeScript contract:
//!  - `Floor` truncates toward -infinity
//!  - `Ceil` rounds toward +infinity (billing uses this so a fractional charge
//!    never shortchanges the merchant)
//!  - `HalfUp` rounds half away from zero
//!  - `HalfEven` rounds half to even (banker's rounding)

use std::fmt;

/// Signed integer unit used by the shared arithmetic. `i128` leaves generous
/// headroom for the scale-12 / scale-24 intermediates used in pricing.
pub type ScaledUnits = i128;

/// Rounding policy applied only when a scaling boundary crosses a remainder.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DecimalRounding {
    Floor,
    Ceil,
    HalfUp,
    HalfEven,
}

impl Default for DecimalRounding {
    fn default() -> Self {
        DecimalRounding::HalfUp
    }
}

/// A parsed decimal value split into integer digits and fractional length, so
/// products / quotients stay exact until a final requested-scale rounding.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct ExactValue {
    digits: ScaledUnits,
    scale: u32,
}

/// Error returned by the exact arithmetic helpers for malformed or out-of
/// `i128`-range input, or an overflowed intermediate.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DecimalMathError {
    /// Input is empty or not a valid `-digits[.digits]` decimal string.
    InvalidInput(String),
    /// An `i128` overflow happened while producing the exact result.
    Overflow,
    /// Division by zero was requested.
    DivideByZero,
}

impl fmt::Display for DecimalMathError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidInput(value) => write!(f, "invalid decimal string: {value:?}"),
            Self::Overflow => f.write_str("decimal arithmetic overflow"),
            Self::DivideByZero => f.write_str("divide by zero in decimal arithmetic"),
        }
    }
}

impl std::error::Error for DecimalMathError {}

type MathResult<T> = Result<T, DecimalMathError>;

fn scale_unit(scale: u32) -> Result<ScaledUnits, DecimalMathError> {
    10i128
        .checked_pow(scale)
        .ok_or(DecimalMathError::Overflow)
}

/// Parse a signed decimal string into the integer count of `scale`-th units.
/// A value carrying more fractional digits than `scale` is rounded to `scale`
/// with `rounding`. Returns `Err` for empty / malformed input or overflow.
pub fn decimal_to_scaled(
    value: &str,
    scale: u32,
    rounding: DecimalRounding,
) -> MathResult<ScaledUnits> {
    let head = value.trim();
    if head.is_empty() {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    let (negative, unsigned) = match head.strip_prefix('-') {
        Some(rest) => (true, rest),
        None => (false, head),
    };
    let mut dot = unsigned.splitn(2, '.');
    let whole_raw = dot.next().unwrap_or("");
    let fraction_raw = dot.next();
    if (unsigned.contains('.') && fraction_raw.is_none()) || whole_raw.is_empty() {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    if !whole_raw.chars().all(|ch| ch.is_ascii_digit()) {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    let fraction = fraction_raw.unwrap_or("");
    if !fraction.chars().all(|ch| ch.is_ascii_digit()) {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    // Exact value as an integer: (whole * 10^fraction_len) + fraction_digits,
    // i.e. the number in units of 10^-fraction_len.
    let n = fraction.len() as u32;
    let whole: ScaledUnits = whole_raw
        .parse()
        .map_err(|_| DecimalMathError::InvalidInput(head.to_owned()))?;
    let mut raw = whole
        .checked_mul(10i128.checked_pow(n).ok_or(DecimalMathError::Overflow)?)
        .ok_or(DecimalMathError::Overflow)?;
    if n > 0 {
        let frac: ScaledUnits = fraction
            .parse()
            .map_err(|_| DecimalMathError::InvalidInput(head.to_owned()))?;
        raw = raw.checked_add(frac).ok_or(DecimalMathError::Overflow)?;
    }
    // Shift into `scale`-N units, rounding only when more digits than `scale`.
    let mut units = if n <= scale {
        raw.checked_mul(10i128.checked_pow(scale - n).ok_or(DecimalMathError::Overflow)?)
            .ok_or(DecimalMathError::Overflow)?
    } else {
        let denom = 10i128
            .checked_pow(n - scale)
            .ok_or(DecimalMathError::Overflow)?;
        round_fraction(raw, denom, rounding)
    };
    if negative {
        units = units.checked_neg().ok_or(DecimalMathError::Overflow)?;
    }
    Ok(units)
}

/// Parse the exact value of a signed decimal string into integer digits and a
/// fractional scale without any loss. Used internally by arithmetic.
fn parse_exact(value: &str) -> MathResult<ExactValue> {
    let head = value.trim();
    if head.is_empty() {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    let (negative, unsigned) = match head.strip_prefix('-') {
        Some(rest) => (true, rest),
        None => (false, head),
    };
    let mut dot = unsigned.splitn(2, '.');
    let whole_raw = dot.next().unwrap_or("");
    let fraction_raw = dot.next();
    if (unsigned.contains('.') && fraction_raw.is_none()) || whole_raw.is_empty() {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    if !whole_raw.chars().all(|ch| ch.is_ascii_digit())
        || fraction_raw.is_some_and(|f| !f.chars().all(|ch| ch.is_ascii_digit()))
    {
        return Err(DecimalMathError::InvalidInput(head.to_owned()));
    }
    let fraction = fraction_raw.unwrap_or("");
    let trimmed = fraction.trim_end_matches('0');
    let whole: ScaledUnits = whole_raw
        .parse()
        .map_err(|_| DecimalMathError::InvalidInput(head.to_owned()))?;
    let mut digits = whole
        .checked_mul(10i128.checked_pow(trimmed.len() as u32).ok_or(DecimalMathError::Overflow)?)
        .ok_or(DecimalMathError::Overflow)?;
    if !trimmed.is_empty() {
        let frac: ScaledUnits = trimmed
            .parse()
            .map_err(|_| DecimalMathError::InvalidInput(head.to_owned()))?;
        digits = digits.checked_add(frac).ok_or(DecimalMathError::Overflow)?;
    }
    if negative {
        digits = digits.checked_neg().ok_or(DecimalMathError::Overflow)?;
    }
    Ok(ExactValue {
        digits,
        scale: trimmed.len() as u32,
    })
}

/// Round `num / denom` to an integer using `rounding` (denom must be positive).
pub fn round_fraction(num: ScaledUnits, denom: ScaledUnits, rounding: DecimalRounding) -> ScaledUnits {
    if denom <= 0 {
        return num;
    }
    let quotient = num / denom;
    let remainder = num % denom;
    if remainder == 0 {
        return quotient;
    }
    let abs_remainder = remainder.unsigned_abs();
    let doubled = abs_remainder.saturating_mul(2);
    let positive = num > 0;
    match rounding {
        DecimalRounding::Ceil => {
            if positive {
                quotient + 1
            } else {
                quotient
            }
        }
        DecimalRounding::Floor => {
            if positive {
                quotient
            } else {
                quotient - 1
            }
        }
        DecimalRounding::HalfEven => {
            if doubled < denom.unsigned_abs() {
                quotient
            } else if doubled > denom.unsigned_abs() {
                if positive {
                    quotient + 1
                } else {
                    quotient - 1
                }
            } else if quotient % 2 == 0 {
                quotient
            } else if positive {
                quotient + 1
            } else {
                quotient - 1
            }
        }
        DecimalRounding::HalfUp => {
            if doubled >= denom.unsigned_abs() {
                if positive {
                    quotient + 1
                } else {
                    quotient - 1
                }
            } else {
                quotient
            }
        }
    }
}

/// Format an integer count of `scale`-th units back into a decimal string,
/// trimming trailing fractional zeros.
pub fn scaled_to_decimal(units: ScaledUnits, scale: u32) -> MathResult<String> {
    let unit = scale_unit(scale)?;
    let negative = units < 0;
    let absolute = units.unsigned_abs();
    let whole = (absolute / unit as u128) as ScaledUnits;
    if scale == 0 {
        return Ok(if negative {
            format!("-{whole}")
        } else {
            whole.to_string()
        });
    }
    let fraction_raw = format!("{:0width$}", absolute % unit as u128, width = scale as usize);
    let fraction = fraction_raw.trim_end_matches('0');
    let sign = if negative { "-" } else { "" };
    if fraction.is_empty() {
        Ok(format!("{sign}{whole}"))
    } else {
        Ok(format!("{sign}{whole}.{fraction}"))
    }
}

fn add_scaled(left: &str, right: &str, scale: u32, rounding: DecimalRounding, negate_right: bool) -> MathResult<String> {
    let unit = scale_unit(scale)?;
    let mut left_units = decimal_to_scaled(left, scale, rounding)?;
    let right_units = decimal_to_scaled(right, scale, rounding)?;
    left_units = if negate_right {
        left_units
            .checked_sub(right_units)
            .ok_or(DecimalMathError::Overflow)?
    } else {
        left_units
            .checked_add(right_units)
            .ok_or(DecimalMathError::Overflow)?
    };
    scaled_to_decimal(left_units, scale)
        .map_err(|_| DecimalMathError::Overflow)
        .map(|v| {
            let _ = unit;
            v
        })
}

/// Exact decimal addition of two strings scaled to `scale`.
pub fn decimal_add(left: &str, right: &str, scale: u32, rounding: DecimalRounding) -> MathResult<String> {
    add_scaled(left, right, scale, rounding, false)
}

/// Exact decimal subtraction of two strings scaled to `scale`.
pub fn decimal_subtract(left: &str, right: &str, scale: u32, rounding: DecimalRounding) -> MathResult<String> {
    add_scaled(left, right, scale, rounding, true)
}

/// Exact decimal multiplication of two strings, returning a decimal string
/// scaled to `scale`. The exact product is computed in integer arithmetic, then
/// rounded to `scale` with `rounding`. Billing calls `(scale 6, Ceil)`.
pub fn decimal_multiply(left: &str, right: &str, scale: u32, rounding: DecimalRounding) -> MathResult<String> {
    let left_exact = parse_exact(left)?;
    let right_exact = parse_exact(right)?;
    let product = left_exact
        .digits
        .checked_mul(right_exact.digits)
        .ok_or(DecimalMathError::Overflow)?;
    let product_scale = left_exact.scale + right_exact.scale;
    let offset = scale as i32 - product_scale as i32;
    let result_units = if offset >= 0 {
        product
            .checked_mul(10i128.checked_pow(offset as u32).ok_or(DecimalMathError::Overflow)?)
            .ok_or(DecimalMathError::Overflow)?
    } else {
        let denom = 10i128
            .checked_pow((-offset) as u32)
            .ok_or(DecimalMathError::Overflow)?;
        round_fraction(product, denom, rounding)
    };
    scaled_to_decimal(result_units, scale)
}

/// Exact decimal division `numerator / denominator`, returning a decimal string
/// scaled to `scale`. Divide-by-zero or malformed input is an error.
pub fn decimal_divide(numerator: &str, denominator: &str, scale: u32, rounding: DecimalRounding) -> MathResult<String> {
    let num_exact = parse_exact(numerator)?;
    let den_exact = parse_exact(denominator)?;
    if den_exact.digits == 0 {
        return Err(DecimalMathError::DivideByZero);
    }
    let num_power = scale as i32 + den_exact.scale as i32 - num_exact.scale as i32;
    let result_units = if num_power >= 0 {
        let scaled =
            num_exact
                .digits
                .checked_mul(10i128.checked_pow(num_power as u32).ok_or(DecimalMathError::Overflow)?)
                .ok_or(DecimalMathError::Overflow)?;
        round_fraction(scaled, den_exact.digits, rounding)
    } else {
        let den_scaled = den_exact
            .digits
            .checked_mul(10i128.checked_pow((-num_power) as u32).ok_or(DecimalMathError::Overflow)?)
            .ok_or(DecimalMathError::Overflow)?;
        round_fraction(num_exact.digits, den_scaled, rounding)
    };
    scaled_to_decimal(result_units, scale)
}

/// Exact summation of decimal strings scaled to `scale` (a sum never needs
/// rounding; malformed entries are counted as zero).
pub fn sum_decimal_strings(values: &[String], scale: u32) -> MathResult<String> {
    let unit = scale_unit(scale)?;
    let mut total = 0i128;
    for value in values {
        if let Ok(units) = decimal_to_scaled(value, scale, DecimalRounding::Floor) {
            total += units;
        }
    }
    scaled_to_decimal(total, scale).or_else(|_| {
        let _ = unit;
        Err(DecimalMathError::Overflow)
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn add(a: &str, b: &str) -> String {
        decimal_add(a, b, 12, DecimalRounding::HalfUp).unwrap()
    }

    fn mul(a: &str, b: &str) -> i128 {
        decimal_to_scaled(
            &decimal_multiply(a, b, 6, DecimalRounding::Ceil).unwrap(),
            6,
            DecimalRounding::Floor,
        )
        .unwrap()
    }

    #[test]
    fn scaled_round_trip_exact() {
        assert_eq!(decimal_to_scaled("1.5", 6, DecimalRounding::Floor).unwrap(), 1_500_000);
        assert_eq!(
            decimal_to_scaled("0.000704", 12, DecimalRounding::Floor).unwrap(),
            704_000_000
        );
        assert_eq!(scaled_to_decimal(1_500_000, 6).unwrap(), "1.5");
        assert_eq!(scaled_to_decimal(704_000_000, 12).unwrap(), "0.000704");
    }

    #[test]
    fn invalid_input_is_error() {
        assert!(decimal_to_scaled("", 6, DecimalRounding::Floor).is_err());
        assert!(decimal_to_scaled("abc", 6, DecimalRounding::Floor).is_err());
        assert!(decimal_to_scaled("1.2.3", 6, DecimalRounding::Floor).is_err());
    }

    #[test]
    fn add_avoids_float_drift() {
        assert_eq!(add("0.1", "0.2"), "0.3");
        assert_eq!(add("10.000004", "0.000001"), "10.000005");
    }

    #[test]
    fn subtract_is_exact() {
        assert_eq!(decimal_subtract("1", "0.000001", 12, DecimalRounding::HalfUp).unwrap(), "0.999999");
        assert_eq!(decimal_subtract("1.00", "1", 12, DecimalRounding::HalfUp).unwrap(), "0");
    }

    #[test]
    fn multiply_converts_cash_times_rate_into_micro_with_ceil() {
        assert_eq!(mul("0.000704", "70"), 49_280);
        assert_eq!(mul("1.23", "6.96"), 8_560_800);
        assert_eq!(mul("0.0000001", "1"), 1);
        assert_eq!(mul("20.00", "75.5"), 1_510_000_000);
    }

    #[test]
    fn divide_computes_scale_12_rate_with_ceil() {
        assert_eq!(
            decimal_divide("0.049280", "0.000704", 12, DecimalRounding::Ceil).unwrap(),
            "70"
        );
        assert_eq!(decimal_divide("1", "3", 6, DecimalRounding::Ceil).unwrap(), "0.333334");
        assert_eq!(decimal_divide("1", "3", 6, DecimalRounding::Floor).unwrap(), "0.333333");
        assert!(decimal_divide("1", "0", 12, DecimalRounding::HalfUp).is_err());
    }

    #[test]
    fn to_scaled_rounds_excess_fraction_digits() {
        assert_eq!(
            decimal_to_scaled("0.12345678", 4, DecimalRounding::Ceil).unwrap(),
            1235
        );
        assert_eq!(
            decimal_to_scaled("0.12345678", 4, DecimalRounding::Floor).unwrap(),
            1234
        );
        assert_eq!(
            decimal_to_scaled("12.9999999", 6, DecimalRounding::Ceil).unwrap(),
            13_000_000
        );
    }

    #[test]
    fn rounding_modes() {
        assert_eq!(decimal_divide("5", "2", 0, DecimalRounding::HalfUp).unwrap(), "3");
        assert_eq!(decimal_divide("5", "2", 0, DecimalRounding::HalfEven).unwrap(), "2");
        assert_eq!(decimal_divide("7", "2", 0, DecimalRounding::HalfEven).unwrap(), "4");
    }

    #[test]
    fn sum_page_costs_exactly() {
        assert_eq!(
            sum_decimal_strings(&["0.000704".into(), "0.000001".into(), "1.5".into()], 6).unwrap(),
            "1.500705"
        );
        assert_eq!(sum_decimal_strings(&[], 6).unwrap(), "0");
    }
}